### ISSUE TITLE
Global executable from path source runs twice on first run on Windows 

### ISSUE THREAD

**Issue by blaugold** [dart-lang/pub#3575]
# Environment

- Dart version (run `dart --version`): 2.18.1
- OS kind and version (e.g. "Windows 10, version 1809" or "macOS 12.4"): Microsoft Windows Server 2022 
- Are you using the Chinese community mirror or a corporate firewall? No

# Problem

When an executable that has been installed with `dart pub global activate --verbose --source path .` is executed for the first time on Windows, it is executed twice.

This [repository](https://github.com/blaugold/pub_global_executable_issue) contains a reproduction with the following executable script and a [GitHub Action workflow](https://github.com/blaugold/pub_global_executable_issue/blob/main/.github/workflows/ci.yaml):

```dart
// bin/foo.dart

import 'dart:io';

void main(List<String> arguments) {
  print('Running foo');
  sleep(Duration(seconds: 5));
}
```

This is the output of running `foo --verbose` for the fist time:

```
$: foo --verbose
FINE: Pub 2.18.1
IO  : Created temp directory D:\a\pub_global_pwsh\pub_global_pwsh\.dart_tool/pub\incremental\foo\tmp340e[8](https://github.com/blaugold/pub_global_pwsh/actions/runs/3122206521/jobs/5063995252#step:5:9)672
MSG : Building package executable...
MSG : Built foo:foo.
IO  : Renaming "D:\a\pub_global_pwsh\pub_global_pwsh\.dart_tool/pub\incremental\foo\tmp340e8672\foo.dart.incremental.dill.incremental.dill" to "D:\a\pub_global_pwsh\pub_global_pwsh\.dart_tool/pub\bin\foo\foo.dart-2.18.1.snapshot".
IO  : Deleting directory D:\a\pub_global_pwsh\pub_global_pwsh\.dart_tool/pub\incremental\foo\tmp340e8672.
FINE: Building package executable finished (0.3[11](https://github.com/blaugold/pub_global_pwsh/actions/runs/3122206521/jobs/5063995252#step:5:12)s).
FINE: Replacing old binstub C:\Users\runneradmin\.pub-cache\bin\foo.bat
IO  : Deleting file C:\Users\runneradmin\.pub-cache\bin\foo.bat.
IO  : Created temp directory C:\Users\runneradmin\.pub-cache\_temp\dir1765ba3e
IO  : Writing 630 characters to text file C:\Users\runneradmin\.pub-cache\bin\foo.bat.
FINE: Contents:
    | @echo off
    | rem This file was created by pub v2.18.1.
    | rem Package: foo
    | rem Version: 0.0.0
    | rem Executable: foo
    | rem Script: foo
    | if exist "D:\a\pub_global_pwsh\pub_global_pwsh\.dart_tool/pub\bin\foo\foo.dart-2.18.1.snapshot" (
    |   call dart "D:\a\pub_global_pwsh\pub_global_pwsh\.dart_tool/pub\bin\foo\foo.dart-2.18.1.snapshot" %*
    |   rem The VM exits with code 253 if the snapshot version is out-of-date.
    |   rem If it is, we need to delete it and run "pub global" manually.
    |   if not errorlevel 253 (
    |     goto error
    |   )
    |   dart pub global run foo:foo %*
    | ) else (
    |   dart pub global run foo:foo %*
    | )
    | goto eof
    | :error
    | exit /b %errorlevel%
    | :eof
    | 
IO  : Deleting directory C:\Users\runneradmin\.pub-cache\_temp\dir[17](https://github.com/blaugold/pub_global_pwsh/actions/runs/3122206521/jobs/5063995252#step:5:18)65ba3e.
Running foo
IO  : Writing [26](https://github.com/blaugold/pub_global_pwsh/actions/runs/3122206521/jobs/5063995252#step:5:27)[40](https://github.com/blaugold/pub_global_pwsh/actions/runs/3122206521/jobs/5063995252#step:5:41) characters to text file C:\Users\runneradmin\.pub-cache\log\pub_log.txt.
MSG : Logs written to C:\Users\runneradmin\.pub-cache\log\pub_log.txt.
The system cannot find the path specified.
Running foo
```

## Expected behavior

The executable is only executed once.

## Actual behavior

The executable is only executed twice.

**Comment by blaugold** [comment]
The original issue that caused me to create a minimal reproduction is somewhat different. There, the first run of the executable fails differently:
```
D:\a\melos\melos>CMD /C melos --verbose bootstrap 
FINE: Pub 2.18.1
MSG : Building package executable...
IO  : Created temp directory D:\a\melos\melos\.dart_tool/pub\incremental\melos_monorepo\tmp4778ffbd
MSG : Built melos_monorepo:melos_dev.
IO  : Renaming "D:\a\melos\melos\.dart_tool/pub\incremental\melos_monorepo\tmp4778ffbd\melos_dev.dart.incremental.dill.incremental.dill" to "D:\a\melos\melos\.dart_tool/pub\bin\melos_monorepo\melos_dev.dart-2.18.1.snapshot".
IO  : Deleting directory D:\a\melos\melos\.dart_tool/pub\incremental\melos_monorepo\tmp4778ffbd.
FINE: Building package executable finished (2.155s).
FINE: Replacing old binstub C:\hostedtoolcache\windows\flutter\stable-3.3.2-x64/.pub-cache\bin\melos.bat
IO  : Deleting file C:\hostedtoolcache\windows\flutter\stable-3.3.2-x64/.pub-cache\bin\melos.bat.
IO  : Created temp directory C:\hostedtoolcache\windows\flutter\stable-3.3.2-x64/.pub-cache\_temp\dir274b2a29
IO  : Writing 677 characters to text file C:\hostedtoolcache\windows\flutter\stable-3.3.2-x64/.pub-cache\bin\melos.bat.
FINE: Contents:
    | @echo off
    | rem This file was created by pub v2.18.1.
    | rem Package: melos_monorepo
    | rem Version: 0.0.0
    | rem Executable: melos
    | rem Script: melos_dev
    | if exist "D:\a\melos\melos\.dart_tool/pub\bin\melos_monorepo\melos_dev.dart-2.18.1.snapshot" (
    |   call dart "D:\a\melos\melos\.dart_tool/pub\bin\melos_monorepo\melos_dev.dart-2.18.1.snapshot" %*
    |   rem The VM exits with code 253 if the snapshot version is out-of-date.
    |   rem If it is, we need to delete it and run "pub global" manually.
    |   if not errorlevel 253 (
    |     goto error
    |   )
    |   dart pub global run melos_monorepo:melos_dev %*
    | ) else (
    |   dart pub global run melos_monorepo:melos_dev %*
    | )
    | goto eof
    | :error
    | exit /b %errorlevel%
    | :eof
    | 
IO  : Deleting directory C:\hostedtoolcache\windows\flutter\stable-3.3.2-x64/.pub-cache\_temp\dir274b2a29.
[     0.193s] melos bootstrap
[     0.204s]   └> D:\a\melos\melos

[     0.205s] Running "dart pub get" in workspace packages...
[    11.273s] [conventional_commit]: Resolving dependencies...
[    11.273s] [conventional_commit]: + _fe_analyzer_shared 48.0.0
[    11.273s] [conventional_commit]: + analyzer 5.0.0
[    11.273s] [conventional_commit]: + args 2.3.1
[    11.273s] [conventional_commit]: + async 2.9.0
[    11.273s] [conventional_commit]: + boolean_selector 2.1.0
[    11.273s] [conventional_commit]: + collection 1.16.0
[    11.273s] [conventional_commit]: + convert 3.0.2
[    11.273s] [conventional_commit]: + coverage 1.6.1
[    11.273s] [conventional_commit]: + crypto 3.0.2
[    11.273s] [conventional_commit]: + file 6.1.4
[    11.273s] [conventional_commit]: + frontend_server_client 3.0.0
[    11.273s] [conventional_commit]: + glob 2.1.0
[    11.274s] [conventional_commit]: + http_multi_server 3.2.1
[    11.274s] [conventional_commit]: + http_parser 4.0.1
[    11.274s] [conventional_commit]: + io 1.0.3
[    11.274s] [conventional_commit]: + js 0.6.4
[    11.274s] [conventional_commit]: + logging 1.0.2
[    11.274s] [conventional_commit]: + matcher 0.12.12
[    11.274s] [conventional_commit]: + meta 1.8.0
[    11.274s] [conventional_commit]: + mime 1.0.2
[    11.274s] [conventional_commit]: + node_preamble 2.0.1
[    11.274s] [conventional_commit]: + package_config 2.1.0
[    11.274s] [conventional_commit]: + path 1.8.2
[    11.274s] [conventional_commit]: + pool 1.5.1
[    11.274s] [conventional_commit]: + pub_semver 2.1.1
[    11.274s] [conventional_commit]: + shelf 1.4.0
[    11.274s] [conventional_commit]: + shelf_packages_handler 3.0.1
[    11.274s] [conventional_commit]: + shelf_static 1.1.1
[    11.274s] [conventional_commit]: + shelf_web_socket 1.0.2
[    11.274s] [conventional_commit]: + source_map_stack_trace 2.1.0
[    11.274s] [conventional_commit]: + source_maps 0.10.10
[    11.274s] [conventional_commit]: + source_span 1.9.1
[    11.275s] [conventional_commit]: + stack_trace 1.10.0
[    11.275s] [conventional_commit]: + stream_channel 2.1.1
[    11.275s] [conventional_commit]: + string_scanner 1.1.1
[    11.275s] [conventional_commit]: + term_glyph 1.2.1
[    11.275s] [conventional_commit]: + test 1.21.6
[    11.275s] [conventional_commit]: + test_api 0.4.14
[    11.275s] [conventional_commit]: + test_core 0.4.18
[    11.275s] [conventional_commit]: + typed_data 1.3.1
[    11.275s] [conventional_commit]: + vm_service 9.4.0
[    11.275s] [conventional_commit]: + watcher 1.0.1
[    11.275s] [conventional_commit]: + web_socket_channel 2.2.0
[    11.275s] [conventional_commit]: + webkit_inspection_protocol 1.2.0
[    11.275s] [conventional_commit]: + yaml 3.1.1
[    11.275s] [conventional_commit]: Downloading test 1.21.6...
[    11.275s] [conventional_commit]: Downloading test_core 0.4.18...
[    11.275s] [conventional_commit]: Downloading test_api 0.4.14...
[    11.275s] [conventional_commit]: Downloading stream_channel 2.1.1...
[    11.275s] [conventional_commit]: Downloading webkit_inspection_protocol 1.2.0...
[    11.275s] [conventional_commit]: Downloading frontend_server_client 3.0.0...
[    11.275s] [conventional_commit]: Downloading shelf 1.4.0...
[    11.275s] [conventional_commit]: Downloading coverage 1.6.1...
[    11.275s] [conventional_commit]: Downloading vm_service 9.4.0...
[    11.275s] [conventional_commit]: Downloading analyzer 5.0.0...
[    11.276s] [conventional_commit]: Downloading _fe_analyzer_shared 48.0.0...
[    11.276s] [conventional_commit]: Changed 45 dependencies!
[    11.316s]   ✓ conventional_commit
[    11.319s]     └> packages/conventional_commit
[    12.199s]   ✓ melos_monorepo
[    12.199s]     └> .
[    12.318s] [melos]: Resolving dependencies...
[    12.318s] [melos]: + _fe_analyzer_shared 48.0.0
[    12.318s] [melos]: + analyzer 5.0.0
[    12.318s] [melos]: + ansi_styles 0.3.2+1
[    12.318s] [melos]: + args 2.3.1
[    12.318s] [melos]: + async 2.9.0
[    12.318s] [melos]: + boolean_selector 2.1.0
[    12.318s] [melos]: + build 2.3.1
[    12.318s] [melos]: + built_collection 5.1.1
[    12.318s] [melos]: + built_value 8.4.1
[    12.318s] [melos]: + charcode 1.3.1
[    12.318s] [melos]: + cli_util 0.3.5
[    12.318s] [melos]: + code_builder 4.3.0
[    12.318s] [melos]: + collection 1.16.0
[    12.318s] [melos]: + convert 3.0.2
[    12.318s] [melos]: + coverage 1.6.1
[    12.318s] [melos]: + crypto 3.0.2
[    12.318s] [melos]: + dart_style 2.2.4
[    12.319s] [melos]: + file 6.1.4
[    12.319s] [melos]: + fixnum 1.0.1
[    12.319s] [melos]: + frontend_server_client 3.0.0
[    12.319s] [melos]: + glob 2.1.0
[    12.319s] [melos]: + graphs 2.1.0
[    12.319s] [melos]: + http 0.13.5
[    12.319s] [melos]: + http_multi_server 3.2.1
[    12.319s] [melos]: + http_parser 4.0.1
[    12.319s] [melos]: + io 1.0.3
[    12.319s] [melos]: + js 0.6.4
[    12.319s] [melos]: + json_annotation 4.7.0
[    12.319s] [melos]: + logging 1.0.2
[    12.319s] [melos]: + matcher 0.12.12
[    12.319s] [melos]: + meta 1.8.0
[    12.319s] [melos]: + mime 1.0.2
[    12.319s] [melos]: + mockito 5.3.2
[    12.319s] [melos]: + mustache_template 2.0.0
[    12.319s] [melos]: + node_preamble 2.0.1
[    12.319s] [melos]: + package_config 2.1.0
[    12.319s] [melos]: + path 1.8.2
[    12.319s] [melos]: + platform 3.1.0
[    12.319s] [melos]: + pool 1.5.1
[    12.319s] [melos]: + process 4.2.4
[    12.319s] [melos]: + prompts 2.0.0
[    12.319s] [melos]: + pub_semver 2.1.1
[    12.319s] [melos]: + pub_updater 0.2.2
[    12.319s] [melos]: + pubspec 2.3.0
[    12.319s] [melos]: + quiver 3.1.0
[    12.319s] [melos]: + shelf 1.4.0
[    12.319s] [melos]: + shelf_packages_handler 3.0.1
[    12.319s] [melos]: + shelf_static 1.1.1
[    12.319s] [melos]: + shelf_web_socket 1.0.2
[    12.319s] [melos]: + source_gen 1.2.5
[    12.319s] [melos]: + source_map_stack_trace 2.1.0
[    12.319s] [melos]: + source_maps 0.10.10
[    12.319s] [melos]: + source_span 1.9.1
[    12.319s] [melos]: + stack_trace 1.10.0
[    12.319s] [melos]: + stream_channel 2.1.1
[    12.319s] [melos]: + string_scanner 1.1.1
[    12.319s] [melos]: + term_glyph 1.2.1
[    12.319s] [melos]: + test 1.21.6
[    12.319s] [melos]: + test_api 0.4.14
[    12.319s] [melos]: + test_core 0.4.18
[    12.319s] [melos]: + typed_data 1.3.1
[    12.319s] [melos]: + uri 1.0.0
[    12.319s] [melos]: + vm_service 9.4.0
[    12.319s] [melos]: + watcher 1.0.1
[    12.320s] [melos]: + web_socket_channel 2.2.0
[    12.320s] [melos]: + webkit_inspection_protocol 1.2.0
[    12.320s] [melos]: + yaml 3.1.1
[    12.320s] [melos]: + yaml_edit 2.0.3
[    12.320s] [melos]: Downloading mockito 5.3.2...
[    12.320s] [melos]: Downloading code_builder 4.3.0...
[    12.320s] [melos]: Downloading analyzer 5.0.0...
[    12.320s] [melos]: Downloading _fe_analyzer_shared 48.0.0...
[    12.320s] [melos]: Downloading build 2.3.1...
[    12.320s] [melos]: Downloading dart_style 2.2.4...
[    12.320s] [melos]: Downloading built_value 8.4.1...
[    12.320s] [melos]: Downloading source_gen 1.2.5...
[    12.320s] [melos]: Downloading test 1.21.6...
[    12.320s] [melos]: Downloading test_api 0.4.14...
[    12.320s] [melos]: Downloading test_core 0.4.18...
[    12.320s] [melos]: Downloading shelf 1.4.0...
[    12.320s] [melos]: Downloading coverage 1.6.1...
[    12.320s] [melos]: Downloading vm_service 9.4.0...
[    12.320s] [melos]: Changed 69 dependencies!
[    12.379s]   ✓ melos
[    12.379s]     └> packages/melos
[    12.381s]   > SUCCESS

[    12.382s] Generating IntelliJ IDE files...
[    12.423s]   > SUCCESS

[    12.423s]  -> 3 packages bootstrapped
IO  : Writing 6410 characters to text file C:\hostedtoolcache\windows\flutter\stable-3.3.2-x64/.pub-cache\log\pub_log.txt.
MSG : Logs written to C:\hostedtoolcache\windows\flutter\stable-3.3.2-x64/.pub-cache\log\pub_log.txt.
The system cannot find the path specified.
Could not find a command named "D:\a\melos\melos\.dart_tool/pub\bin\melos_monorepo\melos_dev.dart-2.18.1.snapshot".

Usage: dart [vm-options] <command|dart-file> [arguments]

Global options:
-h, --help                 Print this usage information.
-v, --verbose              Show additional command output.
    --version              Print the Dart SDK version.
    --enable-analytics     Enable analytics.
    --disable-analytics    Disable analytics.
    --diagnostics          Show tool diagnostic output.

Available commands:
  analyze           Analyze Dart code in a directory.
  compile           Compile Dart to various formats.
  create            Create a new Dart project.
  devtools          Open DevTools (optionally connecting to an existing application).
  doc               Generate API documentation for Dart projects.
  fix               Apply automated fixes to Dart source code.
  format            Idiomatically format Dart source code.
  language-server   Start Dart's analysis server.
  migrate           Perform null safety migration on a project.
  pub               Work with packages.
  run               Run a Dart program.
  test              Run tests for a project.

Run "dart help <command>" for more information about a command.
See https://dart.dev/tools/dart-tool for detailed documentation.
Error: Process completed with exit code 1.
```

It looks to me like here too the executable is executed twice but the second run fails.

**Comment by sigurdm** [comment]
Ouch - I thought we had resolved the "running twice on windows" with https://github.com/dart-lang/pub/pull/3055. Seems there still is an issue.

Maybe the binstub needs to `call` dart in each invocation - not only the first.

**Comment by blaugold** [comment]
The binstub that is executing when running an executable for this first time looks like this:
```bat
@echo off
rem This file was created by pub v2.18.1.
rem Package: foo
rem Version: 0.0.0
rem Executable: foo
rem Script: foo
dart pub global run foo:foo %*
```

It is replaced by a binstub which uses `call dart <snapshot>`. I'm wondering if it's possible that the shell continues reading and executing the second binstub after executing the last line in the first binstub.

**Comment by sigurdm** [comment]
Hmm, I tried to reproduce on a windows machine but failed (tried both powershell and cmd).

But I think you have found the right clue, modifying a batch-file while it is executing seemingly affects the execution!!:

https://stackoverflow.com/questions/906586/changing-a-batch-file-when-its-running

I think what we did by inserting the `call` was just randomly making things work.

Maybe we should just always have a call to the eventual location snapshot, and catch the error if it doesn't exist.
@jonasfj any ideas?
